Implement duck shaping on SymInts

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85808

Duck shaping says that when two input tensors have the same
size, we assume they are symbolically related.  This follows
the same optimization done by inductor.

This optimization is not done completely because we don't
currently install guards corresponding to the duck shape
relationships we created, but overall the guard propagation
for dynamic shape tracing is incomplete at the moment.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Implement duck shaping on SymInts

Duck shaping says that when two input tensors have the same
size, we assume they are symbolically related.  This follows
the same optimization done by inductor.

This optimization is not done completely because we don't
currently install guards corresponding to the duck shape
relationships we created, but overall the guard propagation
for dynamic shape tracing is incomplete at the moment.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -274,17 +277,27 @@ def _get_key(self):
+        # TODO: Put 0/1 specialization in guards
         if val == 0 or val == 1:
             return val
+        # This implements duck-shaping: input sizes that match are assigned
```
This is a tradeoff to speed up tracing as we have less symint to handle at the cost of some potential recompilation?

ezyang replied:
Yep

ezyang replied:
It's also not just tracing speed; it's easier to generate kernels with more constraints like this

ezyang committed:
Update on "Implement duck shaping on SymInts"


Duck shaping says that when two input tensors have the same
size, we assume they are symbolically related.  This follows
the same optimization done by inductor.

This optimization is not done completely because we don't
currently install guards corresponding to the duck shape
relationships we created, but overall the guard propagation
for dynamic shape tracing is incomplete at the moment.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

albanD reviewed (approved):
Sounds good

ezyang committed:
Update on "Implement duck shaping on SymInts"


Duck shaping says that when two input tensors have the same
size, we assume they are symbolically related.  This follows
the same optimization done by inductor.

This optimization is not done completely because we don't
currently install guards corresponding to the duck shape
relationships we created, but overall the guard propagation
for dynamic shape tracing is incomplete at the moment.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -f "ci instability"
